src/util/bip341.rs

```diff
@@ -24,6 +24,29 @@ use std::{error, fmt, io};
 use util::taproot::{TapLeafHash, TapSighashHash};
 use {Script, SigHashType, Transaction, TxOut};
 
+/// Efficientlu calculates signature hash message for legacy, segwit and taproot inputs.
+pub struct SigHashCache<'a> {
+    /// Access to transaction required for various introspection
+    tx: &'a Transaction,
+
+    /// Common cache for taproot and segwit inputs. It's an option because it's not needed for legacy inputs
+    common_cache: Option<CommonCache>,
+
+    /// Cache for taproot v1 inputs
+    taproot_cache: Option<TaprootCache>,
```

Comment:
I think we might as well just always cache everything?

Reply:
As bitcoin core is doing it's more efficient to separate the cache this way so you can compute only what you need and reuse it according to the type of the inputs.

* for legacy: no cache is computed
* for taproot only: no segwit_cache is computed
* for segwit only: no taproot_cache is computed

Even if we don't have the same efficiency requirement as bitcoin core I think it's nice to have (and also @sanket1729 have a preference for that https://github.com/rust-bitcoin/rust-bitcoin/pull/628#issuecomment-878460252)